### ISSUE TITLE
Userspace crashes if VDSO init fails

### ISSUE THREAD

**Issue by mpe** [linuxppc/issues#331]
The VDSO init is written to tolerate failure, we just don't set `vdso_ready` and when processes start we don't set `mm->context.vdso_base`. See eg this comment.

```
	/* vDSO has a problem and was disabled, just don't "enable" it for the
	 * process
	 */
```

But at least with modern userspace this leads to crashes during signal delivery:

```
[  126.291681][ T1380] sh[1380]: segfault (11) at 7fffe0d4dc80 nip 7fffe0d4dc80 lr 7fffe0d4dc80 code 2
[  126.291832][ T1380] sh[1380]: code: e0d4dca0 00007fff 00000000 00000000 00000000 00000000 00000000 00000000
[  126.291936][ T1380] sh[1380]: code: 00000000 00000000 00000000 00000000 <38210080> 380000ac 44000002 00000000
```

Which says we took a SEGV with a permission error (code 2 == SEGV_ACCERR), ie. the address we're trying to execute is not mapped executable.

That's because we took the else path in `handle_rt_signal64()`:

```
	/* Set up to return from userspace. */
	if (vdso64_rt_sigtramp && tsk->mm->context.vdso_base) {
		regs->link = tsk->mm->context.vdso_base + vdso64_rt_sigtramp;
	} else {
		err |= setup_trampoline(__NR_rt_sigreturn, &frame->tramp[0]);
		if (err)
			goto badframe;
		regs->link = (unsigned long) &frame->tramp[0];
	}
```

Which sets up a trampoline on the stack, but the stack is not executable.

Presumably this used to work, either because older userspace made the stack executable by default, or the kernel did it (?).

It's not clear if we should try to fix this by making the stack executable, or if we should just make the VDSO initialisation a requirement to boot, ie. `panic()` if there's some initialisation failure.

**Comment by chleroy** [comment]
Apparently, other architectures (at least x86 and arm64) do the same. But in addition they also take into account a user given trampoline via sa_restorer member of the sigaction struct. Allthough sigaction man page says that that field is obsolete and unspecified by Posix and shouldn't be used, they are supported and used, especially on architectures that do not provide a VDSO sigtramp.

Mapping stack RWX is deprecated nowadays, a recent commit in the kernel even warns about it https://github.com/linuxppc/linux/commit/47a2ebb7f5053387f5753b524f4920b9b829f922

klibc used to set executable stack. Recent work is aiming at making the stack non exec. For sparc and alpha, this was done by providing a sa_restorer to all signal, see https://git.kernel.org/pub/scm/libs/klibc/klibc.git/commit/?id=05350b72cace18417c1ba00c4796a5e19ebb908b and https://git.kernel.org/pub/scm/libs/klibc/klibc.git/commit/?id=570ed1e207cbe38ed487c722f8ac7db68e664a94

For powerpc nothing special was done, see https://git.kernel.org/pub/scm/libs/klibc/klibc.git/commit/?id=27ad55131385821dfe85b0320f4d6ba8861ab7e5 . There is just a comment telling `Kernel uses vDSO for signal return since 2.6.12 (compat) or 2.6.15 (native)`

**Comment by chleroy** [comment]
Whether the stack is executable or not is defined by the PT_GNU_STACK elf header. If this header is not there, in `arch/powerpc/include/asm/page_64.h` we have

```
#define VM_DATA_DEFAULT_FLAGS \
	(is_32bit_task() ? \
	 VM_DATA_DEFAULT_FLAGS32 : VM_DATA_DEFAULT_FLAGS64)

/*
 * This is the default if a program doesn't have a PT_GNU_STACK
 * program header entry. The PPC64 ELF ABI has a non executable stack
 * stack by default, so in the absence of a PT_GNU_STACK program header
 * we turn execute permission off.
 */
#define VM_STACK_DEFAULT_FLAGS32	VM_DATA_FLAGS_EXEC
#define VM_STACK_DEFAULT_FLAGS64	VM_DATA_FLAGS_NON_EXEC

#define VM_STACK_DEFAULT_FLAGS \
	(is_32bit_task() ? \
	 VM_STACK_DEFAULT_FLAGS32 : VM_STACK_DEFAULT_FLAGS64)
```

**Comment by mpe** [comment]
Yeah I realise making the stack executable is a bad idea.

I guess we should just make the VDSO init required, if it fails we panic().

**Comment by chleroy** [comment]
Why not supporting sa_restorer like other architectures ? That way, should the VDSO not be there, the userland can still decide on either setting an executable stack or giving an sa_restorer.

What if an app unmaps the VDSO ?

Do statically linked apps maps the VDSO ? At least it seems that libc doesn't use gettimeofday from VDSO with statically linked apps on PPC32 at least. Don't know about signal trampolines.

**Comment by chleroy** [comment]
What are the reason that could make the VDSO init fail:
1/ Lack of memory. In that case, I guess will have a lot more problems than that.
2/ Required Symbols not found. Other architectures seem to resolves those symbols at build time. In ARM64 there is a #include <generated/vdso-offsets.h>, and nothing similar to what PPC do at VDSO init. Could powerpc do the same ? In that case, no much other opportunity for the VDSO init to fail.

**Comment by mpe** [comment]
We actually claim `#define __ARCH_HAS_SA_RESTORER` in `arch/powerpc/include/asm/signal.h`.

So seemingly we should support `SA_RESTORER`, I wonder if we did once but broke it?

Statically linked apps do use the VDSO, and the kernel will use it for sigreturn.

**Comment by mpe** [comment]
Yeah I agree VDSO init failing is pretty unlikely, so maybe we just don't worry about it.

I just stumbled across it because I'm trying to get the C VDSO working on ppc64 and having trouble :-)

**Comment by chleroy** [comment]
I submitted a series https://patchwork.ozlabs.org/project/linuxppc-dev/list/?series=204396

I changed the BUG_ON to panic() when memory allocation fails. Anyway there is not much to do if that happens at startup, I can't imagine reaching `init` anyway if memory allocation start failing that early.

Other reason to fail were the required symbols not found. My series implement buildtime symbol fetching instead of runtime fetching, leading to a faillessly startup once the memory allocation succeeded, so the case VDSO init fails is now gone.
